reinforcement_learning/rlclientlib/ranking_event.cc

```diff
@@ -51,100 +50,127 @@ namespace reinforcement_learning {
   ranking_event::ranking_event()
   { }
 
-  ranking_event::ranking_event(const char* event_id, float pass_prob, const std::string& body)
+  ranking_event::ranking_event(const char* event_id, float pass_prob, const char* context, const ranking_response& response)
     : event(event_id, pass_prob)
-    , _body(body)
-  { }
-
+    , _context(context)
+    , _model_id(response.get_model_id())
+  {
+    for (auto const &r : response) {
+      _a_vector.push_back(r.action_id);
+      _p_vector.push_back(r.probability);
+    }
+  }
+  
   ranking_event::ranking_event(ranking_event&& other)
     : event(std::move(other))
-    , _body(std::move(other._body))
-  {}
+    , _context(std::move(other._context))
+    , _a_vector(std::move(other._a_vector))
+    , _p_vector(std::move(other._p_vector))
+    , _model_id(std::move(other._model_id))
+  { }
 
   ranking_event& ranking_event::operator=(ranking_event&& other) {
     if (&other != this) {
       event::operator=(std::move(other));
-      _body = std::move(other._body);
+      _context = std::move(other._context);
+      _a_vector = std::move(other._a_vector);
+      _p_vector = std::move(other._p_vector);
+      _model_id = std::move(other._model_id);
     }
     return *this;
   }
 
-  void ranking_event::serialize(u::data_buffer& oss) {
-    oss << _body;
-    if (_pass_prob < 1) {
-      oss << R"(,"pdrop":)" << (1 - _pass_prob);
-    }
-    oss << R"(})";
+  flatbuffers::Offset<RankingEvent> ranking_event::serialize_eventhub_message(flatbuffers::FlatBufferBuilder& builder) {
```

Comment:
I am not convinced we need to tie the implementation of our events to flatbuffers so tightly. Can this be a static method elsewhere or a different class?

Reply:
It needs to access all fields in events. If we move it to anther class or static method, it will break the encapsulation of events anyway. In this sense, I think staying here should be fine.

Reply:
The data should be able to be accessed externally from ranking_event. It is essentially a data struct. I don't think it is acceptable leaving this in the class.